From d0446a6f8d126fe551587a8575f8625a93f3d5ae Mon Sep 17 00:00:00 2001
From: Shaoyun Liu <Shaoyun.Liu@amd.com>
Date: Mon, 10 Apr 2017 15:33:40 -0400
Subject: [PATCH 1659/4131] adm/amdgpu: Fix PPLIB NULL function pointer when
 load amdkfd on SRIOV VF

On SRIOV VF, the PPLIB will not be enabled and hence powerplay.pp_funcs will
not be initialized which will cause kernel panic when KFD driver call back
to amdgpu for any power play related functoins.

Change-Id: Ifd1e02c2f0530f1f3b9d7cbc18a35dc13f6aaca6
Signed-off-by: Shaoyun Liu <Shaoyun.Liu@amd.com>

 Conflicts:
	drivers/gpu/drm/amd/amdgpu/amdgpu_amdkfd.c
	drivers/gpu/drm/amd/amdgpu/amdgpu_vm.c
---
 drivers/gpu/drm/amd/amdgpu/amdgpu_amdkfd.c | 12 +++++++++---
 drivers/gpu/drm/amd/amdgpu/amdgpu_vm.c     |  2 +-
 2 files changed, 10 insertions(+), 4 deletions(-)

diff --git a/drivers/gpu/drm/amd/amdgpu/amdgpu_amdkfd.c b/drivers/gpu/drm/amd/amdgpu/amdgpu_amdkfd.c
index 286b724..30e5893 100755
--- a/drivers/gpu/drm/amd/amdgpu/amdgpu_amdkfd.c
+++ b/drivers/gpu/drm/amd/amdgpu/amdgpu_amdkfd.c
@@ -368,7 +368,10 @@ void get_local_mem_info(struct kgd_dev *kgd,
 			mem_info->local_mem_size_public,
 			mem_info->local_mem_size_private);
 
-	mem_info->mem_clk_max = amdgpu_dpm_get_mclk(adev, false) / 100;
+	if (amdgpu_sriov_vf(adev))
+		mem_info->mem_clk_max = adev->clock.default_mclk / 100;
+	else
+		mem_info->mem_clk_max = amdgpu_dpm_get_mclk(adev, false) / 100;
 }
 
 uint64_t get_gpu_clock_counter(struct kgd_dev *kgd)
@@ -383,8 +386,11 @@ uint64_t get_gpu_clock_counter(struct kgd_dev *kgd)
 uint32_t get_max_engine_clock_in_mhz(struct kgd_dev *kgd)
 {
 	struct amdgpu_device *adev = (struct amdgpu_device *)kgd;
-        /* The sclk is in quantas of 10kHz */
-	return amdgpu_dpm_get_sclk(adev, false) / 100;
+        if (amdgpu_sriov_vf(adev))
+                return adev->clock.default_sclk / 100;
+        else
+                /* The sclk is in quantas of 10kHz */
+                return amdgpu_dpm_get_sclk(adev, false) / 100;
 }
 
 void get_cu_info(struct kgd_dev *kgd, struct kfd_cu_info *cu_info)
diff --git a/drivers/gpu/drm/amd/amdgpu/amdgpu_vm.c b/drivers/gpu/drm/amd/amdgpu/amdgpu_vm.c
index 2b24bbb..43fba1a 100644
--- a/drivers/gpu/drm/amd/amdgpu/amdgpu_vm.c
+++ b/drivers/gpu/drm/amd/amdgpu/amdgpu_vm.c
@@ -2606,7 +2606,7 @@ int amdgpu_vm_init(struct amdgpu_device *adev, struct amdgpu_vm *vm,
 
 		mutex_lock(&id_mgr->lock);
 
-		if (adev->vm_manager.n_compute_vms++ == 0) {
+		if ((!amdgpu_sriov_vf(adev)) && adev->vm_manager.n_compute_vms++ == 0) {
 			/* First Compute VM: enable compute power profile */
 			if (adev->pp_enabled)
 				amdgpu_dpm_switch_power_profile(adev,
-- 
2.7.4

